### PR TITLE
Extended diagnostic typings

### PR DESCRIPTION
**Pull request by Uzlopak**
Based on #52 but also provides fancier typings. 

Separated it from the nodenext PR, as this is just something fancy, which probably results in discussions...

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1,25 +1,57 @@
...
+    request: FastifyRequest;
+  }
+
+  
+  export interface OnRequestEvent {
+    reply: FastifyReply;
+    request: FastifyRequest;
+  }
+  /**
+   * @deprecated Use OnRequestEvent instead
```
Comment (RafaelGSS) [inline types/index.d.ts:47]: I'm not sure I understand. You are deprecating and recommending the deprecated usage?

File: types/index.d.ts [hunk 2]
```diff
@@ -1,25 +1,57 @@
...
+declare module 'diagnostics_channel' {
+  type FastifyChannelListener<N, T> = (message: T, name: N) => void;
 
-export declare interface OnErrorEvent {
-  reply: FastifyReply;
-  request: FastifyRequest;
-  error: Error;
-}
+  // @ts-ignore Enforce extending from private Class Channel
+  class FastifyChannel<N extends string, T> extends Channel {
```
Comment (RafaelGSS) [inline types/index.d.ts:9]: It should be an interface instead.

**Review by RafaelGSS — COMMENTED** [review]
Technically this module is not needed anymore. IIRC We've implemented most of those messages directly on fastify.

File: types/index.d.ts (hunk 1 above)
Comment (climba03003) [reply types/index.d.ts:47]: `o` vs `O`

**Comment by jsumners** [comment]
Please do not request TypeScript reviews from me.

**Comment by Uzlopak** [comment]
@RafaelGSS 

We could also archive this repository and deprecate it on npm.

**Comment by RafaelGSS** [comment]
> @RafaelGSS
> 
> We could also archive this repository and deprecate it on npm.

SGTM, we should just make sure my statement is correct.

**Pull request closed without merging**